upath/core.py

```diff
@@ -432,24 +451,23 @@ def is_absolute(self) -> bool:
     def unlink(self, missing_ok: bool = False) -> None:
         if not self.exists():
             if not missing_ok:
-                raise FileNotFoundError
-            else:
-                return
+                raise FileNotFoundError(str(self))
+            return
         self._accessor.rm(self, recursive=False)
 
     def rmdir(self, recursive: bool = True) -> None:
-        """Add warning if directory not empty
-        assert is_dir?
-        """
         if not self.is_dir():
-            raise NotADirectoryError
+            raise NotADirectoryError(str(self))
+        if not recursive and next(self.iterdir()):  # type: ignore
+            raise OSError(f"Not recursive and directory not empty: {self}")
         self._accessor.rm(self, recursive=recursive)
 
     def chmod(self, mode, *, follow_symlinks: bool = True) -> None:
         raise NotImplementedError
 
-    def rename(self, target):
+    def rename(self, target, **kwargs):
```

Comment:
why extend the function signature here?

`pathlib.Path.rename` https://github.com/python/cpython/blob/86aa8a5e98e5742bb5bb6ab617f08b1679b33b55/Lib/pathlib.py#L1053

Reply:
It was intended to allow for the keyword arguments that fsspec rename/mv can take, but I just went ahead and added a simple implementation and tests.